### ISSUE TITLE
Proposal :rewrite registry for namespace module

### ISSUE THREAD

**Issue by axetroy** [denoland/registry#117]
### Current problems

1. Namespace conflict
2. Existing packages are no longer maintained
3. Each package needs to initiate a PR merge for [deno/registry](https://github.com/denoland/registry), It also requires staff review and merges

#### Solution 1: use `@username/library` instead of `library` #26

I don't think it's a good idea. This will look ugly like this.

and problem 3 still exists

```typescript
import { echo } from 'https://deno.land/x/@username/echo@v1.0.0/mod.ts';
```

#### Solution 2: using `Golang` style package manage

Learning from Golang package manage.

```go
import (
	"github.com/user1/log"
    "github.com/user2/db"
)
```

We can do this in Deno

```typescript
import log from 'https://deno.land/github.com/user1/log/mod.ts';
import db from 'https://deno.land/github.com/user2/db/mod.ts';

// or with the version tag
import log from 'https://deno.land/github.com/user1/log@v1.0.0/mod.ts';
import db from 'https://deno.land/github.com/user2/db@v1.0.0/mod.ts';
```

This solution resolves all problems.

**No central storage, no need to review a new package's PR**

Can even import from a different code hosting platform. eg `gitlab`

### Try it out

Here is an example I write to show how it works.

https://github.com/axetroy/deno_registry

**Comment by ry** [comment]
You make a good argument - and it would be great to have the registry in deno instead of node.

Do you think you could get that code base to be feature complete and backwards compatible with the current registry? If so I would be happy to invest the time to switch deployments.

**Comment by axetroy** [comment]
Unfortunately, I am very busy at work recently.

If anyone is interested, you can try.

**Comment by bartlomieju** [comment]
I can take a look at it at the end of week

**Comment by j-f1** [comment]
Here’s how package names work for the top 20 languages on GitHub ([list source](https://github.com/oprogramador/github-languages#most-popular-languages-on-github-as-of-february-18-2019))

- bare identifiers: <sup>1</sup>JS [npm], <sup>4</sup>Python [[PyPI](https://pypi.org)], <sup>7</sup>Ruby [[RubyGems](https://rubygems.org/gems)], <sup>8</sup>C# [[NuGet](https://www.nuget.org/packages)], <sup>14</sup>Objective-C/<sup>15</sup>Swift [[CocoaPods](https://guides.cocoapods.org/using/using-cocoapods.html)], <sup>20</sup>Perl [[PPM](http://code.activestate.com/ppm/)]
- scoped identifiers: <sup>1</sup>JS [npm], <sup>2</sup>Java/<sup>19</sup>Scala [[Maven](https://search.maven.org/search?q=a:base64)/[Gradle](https://docs.gradle.org/current/userguide/dependency_types.html)], <sup>5</sup>PHP [[Composer + Packagist](https://packagist.org)]
- arbitrary URLs: <sup>1</sup>JS [npm], <sup>4</sup>Python [[PyPI](https://packaging.python.org/tutorials/installing-packages/#installing-from-vcs)], <sup>5</sup>PHP [[Composer](https://getcomposer.org/doc/02-libraries.md#publishing-to-a-vcs)], Go
- N/A: <sup>3</sup>HTML, <sup>6</sup>CSS, <sup>10</sup>C/<sup>9</sup>C++ [source: [Reddit](https://www.reddit.com/r/cpp/comments/3d1vjq/is_there_a_c_package_manager_if_not_how_do_you/)], <sup>11</sup>Shell
- ???: <sup>17</sup>R, <sup>18</sup>MATLAB

**Comment by axetroy** [comment]
@bartlomieju Are you interested in this?

This may require 3 jobs

- [ ] Rewrite with `Deno`
- [ ] Compatible with the current package
- [ ] Write new package manager

Based on my second solution above.

Is compatible with the old package management method.

eg this package.

https://github.com/denoland/registry/blob/7b90fe8aec7aa11bf94bedc7570b2f49760f4d97/src/database.json#L1-L5

### Old

```typescript
import { server } from 'https://deno.land/x/abc/mod.ts';
```

### New

```typescript
import { server } from `https://deno.land/github.com/zhmushan/abc/mod.ts`;

// Compatible with old packages
// Query whether `/x/:name/:filepath` is defined in `database.json`
// if exist. redirect to it's repo
import { server } from 'https://deno.land/x/abc/mod.ts';
```

If the new package manager land.

No longer need to submit a PR to add to `database.json`

**Comment by bartlomieju** [comment]
@axetroy I may have to put it off, I'm still working on rewrite of DenoDir

**Comment by axetroy** [comment]
Report progress: currently compatible with the old package in [axetroy/deno_registry](https://githb.com/axetroy/deno_registry).

You can import like this:

```ts
import { server } from "http://localhost/x/abc/mod.ts";
import * as math from "http://localhost/x/math/mod.ts";

// or use the new way
import { server } from "http://localhost/github.com/zhmushan/abc/mod.ts";
import * as math from "http://localhost/github.com/axetroy/deno_math/mod.ts";
```

Welcome to share your idea.

**Comment by ry** [comment]
@axetroy, I wonder if you could reuse the code here somehow. @j-f1 has done some amazing work getting the source code displayed properly on browsers and other features. As much as I'd like registry to run on Deno, it's going to need to be close to feature-complete with the current site.

**Comment by j-f1** [comment]
I’m wondering — what do we do with the current bare-identifier-named modules? Should we just throw an error telling people to migrate to the new import URL? It seems risky to “bless” some third-party modules with bare-identifier status simply due to their age.

**Comment by ry** [comment]
> what do we do with the current bare-identifier-named modules?

I’m not sure what you mean. Can you give an example?

**Comment by j-f1** [comment]
https://deno.land/x/abc/README.md — it’s a third-party library with a bare name (`abc`). What will happen to it when/if the registry switches to namespaced names?

**Comment by axetroy** [comment]
> [deno.land/x/abc/README.md](https://deno.land/x/abc/README.md) — it’s a third-party library with a bare name (`abc`). What will happen to it when/if the registry switches to namespaced names?

it should be backward compatible, the original usage is unchanged

and ``database.json`` will not accept a new module.

Existing modules will be marked as deprecated in the documentation

**Comment by tunnckoCore** [comment]
I don't think it should be backward compatible. Deno is in pretty early stages so now is the time to break a bit from time to time. Also, there are not a ton of users yet (not sure?) and migration isn't that hard.

Just drop the database thing and rely on the new way, completely.

**Comment by axetroy** [comment]
I deployed an online example on `heroku`.

try it out online.

[std/http/server.ts](https://denolib.herokuapp.com/std/http/server.ts)
[x/math/mod.ts](https://denolib.herokuapp.com/x/math/mod.ts)
[github.com/zhmushan/abc/mod.ts](https://denolib.herokuapp.com/github.com/zhmushan/abc/mod.ts)
[github.com/chiefbiiko/base64/mod.ts](https://denolib.herokuapp.com/github.com/chiefbiiko/base64/mod.ts)

Now when I encounter problems, the registry only proxy `deno.land/x`

And `deno.land` is deployed on [denoland/deno/website](https://github.com/denoland/deno/tree/master/website), which will cause trouble

Should we consider separating the website into a single repo `denoland/website`

**Comment by j-f1** [comment]
Alternately, the `deno.land` domain could be the registry, with the actual website content on `www.deno.land` and a redirect from https://deno.land → https://www.deno.land for convenience.
